cpp/src/copying/scatter.cu

```diff
@@ -442,6 +444,35 @@ std::vector<std::unique_ptr<table>> scatter_to_tables(
     input, partition_map, mr, stream);
 }
 
+std::unique_ptr<column> boolean_mask_scatter(
+    column_view const& source, column_view const& target,
+    column_view const& boolean_mask,
+    rmm::mr::device_memory_resource *mr,
+    cudaStream_t stream) {
+
+    CUDF_EXPECTS(boolean_mask.size() == target.size(), "Boolean mask size mismatch with traget size");
+    CUDF_EXPECTS(boolean_mask.type().id() == BOOL8, "Mask must be Boolean type");
+    CUDF_EXPECTS(source.type().id() == target.type().id(), "Type mismatch for source and target");
+
+    auto indices = cudf::make_numeric_column(data_type{INT32},
+                                  target.size(), UNALLOCATED, stream, mr);
+    auto mutable_indices = indices->mutable_view();
+
+    thrust::sequence(rmm::exec_policy(stream)->on(stream),
+                     mutable_indices.begin<size_type>(),
+                     mutable_indices.end<size_type>(),
+                     0);
+
+    // The scatter map is actually a table with only one column, which is scatter map.
+    auto scatter_map = detail::apply_boolean_mask(table_view{{indices->view()}}, boolean_mask,
+                                                  mr, stream);
```

Comment:
Suggesting this algorithm of creating `scatter_map` from `boolean_mask` to reduce memory usage and memory writes & reads. Also adds check on number of true's in `boolean_mask`.


```suggestion
auto num_true=thrust::count(rmm::exec_policy(stream)->on(stream),
              boolean_mask.begin<bool>(), boolean_mask.end<bool>(), true);
CUDF_EXPECTS(num_true == source.size(), "Number of true's in mask must be equal to source size");
auto scatter_map = cudf::make_numeric_column(data_type{INT32},
                  source.size(), UNALLOCATED, stream, mr);
auto mutable_scatter_map = scatter_map->mutable_view();
thrust::copy_if(rmm::exec_policy(stream)->on(stream),
                thrust::counting_iterator<size_type>(0),
                thrust::counting_iterator<size_type>(boolean_mask.size()), 
                boolean_mask.begin<bool>(), mutable_scatter_map.begin<size_type>(), 
                thrust::identity<bool>{});
```

Add stream to thrust calls..

Reply:
This is considering the boolean_mask will always be non-null column, I am also not sure of nay use case where it can be null, @kkraus14 do you have any suggestions ?

Reply:
We should follow the behavior of boolean mask gathers here and the two should remain consistent. I believe boolean mask gather currently treats nulls as `False` so we should do the same here.